tt_metal/impl/dispatch/command_queue.cpp

```diff
@@ -1165,6 +1188,151 @@ uint32_t Trace::instantiate(CommandQueue& cq) {
+    auto runtime_args_md = RuntimeArgsMetadata {
+            .core_coord = core_coord,
+            .runtime_args_vec = runtime_args_vec,
+            .kernel = kernel,
+            .update_idx = update_idx,
+    };
+    cq.run_command( CommandInterface {
+        .type = EnqueueCommandType::UPDATE_RUNTIME_ARGS,
+        .blocking = blocking,
+        .runtime_args_md = runtime_args_md,
+    });
+}
+
+void EnqueueUpdateRuntimeArgsImpl (const RuntimeArgsMetadata& runtime_args_md) {
+    std::vector<uint32_t> resolved_runtime_args = {};
+    resolved_runtime_args.reserve((*runtime_args_md.runtime_args_vec).size());
+
+    for (const auto& arg : *(runtime_args_md.runtime_args_vec)) {
+        std::visit([&resolved_runtime_args] (auto&& a) {
+            using T = std::decay_t<decltype(a)>;
+            if constexpr (std::is_same_v<T, Buffer*>) {
+                resolved_runtime_args.push_back(a -> address());
+            } else {
+                resolved_runtime_args.push_back(a);
+            }
+        }, arg);
+    }
+    auto& kernel_runtime_args = runtime_args_md.kernel->runtime_args(runtime_args_md.core_coord);
+    for (const auto& idx : runtime_args_md.update_idx) {
+        kernel_runtime_args[idx] = resolved_runtime_args[idx];
+    }
+}
+
+void EnqueueSetRuntimeArgs(CommandQueue& cq, const std::shared_ptr<Kernel> kernel, const CoreCoord &core_coord, std::shared_ptr<RuntimeArgs> runtime_args_vec, bool blocking) {
+    auto runtime_args_md = RuntimeArgsMetadata {
+            .core_coord = core_coord,
+            .runtime_args_vec = runtime_args_vec,
+            .kernel = kernel,
+    };
+    cq.run_command( CommandInterface {
+        .type = EnqueueCommandType::SET_RUNTIME_ARGS,
+        .blocking = blocking,
+        .runtime_args_md = runtime_args_md,
+    });
+}
+
+void EnqueueSetRuntimeArgsImpl(const RuntimeArgsMetadata& runtime_args_md) {
+    std::vector<uint32_t> resolved_runtime_args = {};
+    resolved_runtime_args.reserve((*runtime_args_md.runtime_args_vec).size());
+
+    for (const auto& arg : *(runtime_args_md.runtime_args_vec)) {
+        std::visit([&resolved_runtime_args] (auto&& a) {
+            using T = std::decay_t<decltype(a)>;
+            if constexpr (std::is_same_v<T, Buffer*>) {
+                resolved_runtime_args.push_back(a -> address());
+            } else {
+                resolved_runtime_args.push_back(a);
+            }
+        }, arg);
+    }
+    runtime_args_md.kernel -> set_runtime_args(runtime_args_md.core_coord, resolved_runtime_args);
+}
+
+void EnqueueGetBufferAddr(CommandQueue& cq, uint32_t* dst_buf_addr, const Buffer* buffer, bool blocking) {
+    cq.run_command( CommandInterface {
+        .type = EnqueueCommandType::GET_BUF_ADDR,
+        .blocking = blocking,
+        .shadow_buffer = buffer,
+        .dst = dst_buf_addr
+    });
+}
+
+void EnqueueGetBufferAddrImpl(void* dst_buf_addr, const Buffer* buffer) {
+    *(static_cast<uint32_t*>(dst_buf_addr)) = buffer -> address();
```

Comment:
I find this API a bit strange. Do we need to pass a pointer around? Can't we just return the buffer's address?

Reply:
Unfortunately, couldn't find a way around it. The value stored at `dst_buf_addr` needs to be asynchronously updated. Returning the buffer address would require blocking until the `get` operation is completed in the CQ.

This way, we "assign" a variable to a buffer address asynchronously and expect it to be updated some time in the future.